### ISSUE TITLE
`equations` silently changes labels – better: type-checked labels

### ISSUE THREAD

**Issue by jorsn** [leftaroundabout/Symbolic-math-HaTeX#2]
The function `equations` creates labels using

```
asSafeLabel = LaTeX.label . fromString . filter isAlpha
```

defined at 
<https:../blob/dbfff7d6ab0f184892e62d3c9829495fa46a170f/Math/LaTeX/Internal/Display.hs#L114>.

`filter isAlpha` silently changes labels named e.g. like `eq:…`, hence breaks common practices of LaTeX label naming.

Since this filtering is not necessarily expected or desired by a user (like me), I'd suggest to make it optional (e.g. provide a version of `equations` with an argument `filterLabel :: String -> String`).

Really nice would be compile-time checking of labels and failure if it didn't fit the specification. I have implemented something as proof of concept:
https://gist.github.com/jorsn/b86dd25b73b6b3bd6685a5a403bc481f

Should this better go into HaTeX directly? It can probably also be modified to sanitize general HaTeX labels.

**Comment by leftaroundabout** [comment]
Absolutely. That seems to be a hack I wrote whilst in a hurry to get some paper compiling, it is not sane by any means.

Your work on type-checked labels is really cool! _Maybe_ this would make sense to include in HaTeX, but HaTeX is supposed to be relatively stable and simple; I've thought a lot about how stuff like references could be made more Haskell-idiomatic, but frankly most of those ideas turn out to be arcane and impractical for actual documents.

But, I'll gladly accept such feature into `TeX-my-math`, and if it turns out to work nicely we can discuss over at HaTeX whether it would be worth porting them there.

**Comment by jorsn** [comment]
Which properties/abilities do you think the label handling should have? I imagine

1. checking whether a new label aready exists,
1. checking whether a referenced label is defined,
1. maybe sanitizing labels like in the gist,
1. maybe listing all labels a function generates (like in `EqnT labels`),
1. having nice, compact syntax.

It would be cool if all this would happen during type-checking, but more pragmatic (easier, performant string splitting) would probably be to just use a fixed point of a recursion hidden in a reader and writer monad, so a function type could be
```haskell
MonadWriter m Labels, MonadReader m Labels, LaTeXC (m a) => m a,
```
plus exceptions/Either/…, where `Labels` is the type of the set of defined labels. The LaTeX compiling stage is there anyway, which can fail and does so sometimes.

You have probably thought about this. What is arcane or impractical about it?
